ml-algorithms/src/main/java/org/opensearch/ml/engine/ingest/OpenAIDataIngestion.java

```diff
@@ -0,0 +1,174 @@
+        } catch (PrivilegedActionException e) {
+            throw new RuntimeException("Failed to read from OpenAI file API: ", e);
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            throw new OpenSearchStatusException("Failed to batch ingest: " + e.getMessage(), RestStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        return successRate;
+    }
+
+    private ActionListener<BulkResponse> getBulkResponseListener(
+        AtomicInteger successfulBatches,
+        AtomicInteger failedBatches,
+        CompletableFuture<Void> future
+    ) {
+        return ActionListener.wrap(bulkResponse -> {
+            if (bulkResponse.hasFailures()) {
+                failedBatches.incrementAndGet();
+                future.completeExceptionally(new RuntimeException(bulkResponse.buildFailureMessage()));  // Mark the future as completed
+                // with an exception
+            }
+            log.debug("Batch Ingestion successfully");
+            successfulBatches.incrementAndGet();
+            future.complete(null); // Mark the future as completed successfully
+        }, e -> {
+            log.error("Failed to bulk update model state", e);
+            failedBatches.incrementAndGet();
+            future.completeExceptionally(e);  // Mark the future as completed with an exception
+        });
+    }
+
+    private void batchIngest(
+        List<String> sourceLines,
+        MLBatchIngestionInput mlBatchIngestionInput,
+        ActionListener<BulkResponse> bulkResponseListener
+    ) {
+        BulkRequest bulkRequest = new BulkRequest();
+        sourceLines.stream().forEach(jsonStr -> {
+            Map<String, Object> jsonMap = processFieldMapping(jsonStr, mlBatchIngestionInput.getFieldMapping());
+            IndexRequest indexRequest = new IndexRequest(mlBatchIngestionInput.getIndexName()).source(jsonMap);
+
+            bulkRequest.add(indexRequest);
+        });
+        client.bulk(bulkRequest, bulkResponseListener);
+    }
+
+    private Map<String, Object> processFieldMapping(String jsonStr, Map<String, Object> fieldMapping) {
+        String outputJsonPath = (String) fieldMapping.get(OUTPUT);
+        List<List> outputs = (List<List>) JsonPath.read(jsonStr, outputJsonPath);
+        List<String> outputFields = (List<String>) fieldMapping.get(OUTPUTIELDS);
+        List<String> ingestFieldsJsonPath = (List<String>) fieldMapping.get(INGESTFIELDS);
+
+        Map<String, Object> jsonMap = new HashMap<>();
+        if (outputs.size() != outputFields.size()) {
+            throw new IllegalArgumentException("the fieldMapping and source data do not match");
+        }
+        for (int index = 0; index < outputs.size(); index++) {
+            jsonMap.put(outputFields.get(index), outputs.get(index));
+        }
+
+        for (String fieldPath : ingestFieldsJsonPath) {
+            jsonMap.put(obtainFieldNameFromJsonPath(fieldPath), JsonPath.read(jsonStr, fieldPath));
```

Comment:
Why need to call `obtainFieldNameFromJsonPath` here? Can you explain with some example?

Reply:
This is to fetch the field name from the JsonPath. For example, in the openAI JsonPath "$.custom_field" retrieve the data from that jsonPath and save it in the field name "custom_field" in the target index.